components/browser/domains/src/main/java/mozilla/components/domains/DomainAutoCompleteProvider.kt

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.domains
+
+import android.content.Context
+import kotlinx.coroutines.experimental.CommonPool
+import kotlinx.coroutines.experimental.android.UI
+import kotlinx.coroutines.experimental.async
+import kotlinx.coroutines.experimental.launch
+import java.util.Locale
+
+typealias ResultCallback = (String, String, Int) -> Unit
+
+/**
+ * Provides autocomplete functionality for domains, based on a provided list
+ * of assets (see @{link Domains}) and/or a custom domain list managed
+ * by {@link CustomDomains}.
+ */
+class DomainAutoCompleteProvider {
+    object AutocompleteSource {
+        const val DEFAULT_LIST = "default"
+        const val CUSTOM_LIST = "custom"
+    }
+
+    private var customDomains: List<String> = emptyList()
+    private var shippedDomains: List<String> = emptyList()
+    private var useCustomDomains: Boolean = false
+    private var useShippedDomains: Boolean = true
```

Comment:
Technically there's no need to define the type if you assign a value. The compiler will do that automatically. E.g. this could be:

```Kotlin
    private var customDomains = emptyList<String>()
    private var shippedDomains = emptyList<String>()
    private var useCustomDomains = false
    private var useShippedDomains = true
```